cross_validation_mpi.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+#import multiprocessing as mp
+import os
+import pathlib as pl
+import pandas as pd
+#from d3m_profiler.build_table import get_datasets
+#import json
+import time
+import pickle
+import sys
+from mpi4py import MPI
+from imblearn.over_sampling import SMOTE
+from imblearn.over_sampling import BorderlineSMOTE, ADASYN
+from sklearn.model_selection import LeaveOneGroupOut
+from sklearn.metrics import confusion_matrix
+from sklearn.metrics import accuracy_score, f1_score
+from sklearn.ensemble import RandomForestClassifier as RandomForestClassifier
+from sklearn.decomposition import PCA
+from sklearn.pipeline import Pipeline
+#from sklearn.neural_network import MLPClassifier as MLPClassifier
+#from sklearn.ensemble import AdaBoostClassifier as AdaBoostClassifier
+#from sklearn.ensemble import GradientBoostingClassifier
+#from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
+#from sklearn.naive_bayes import GaussianNB as GaussianNB
+#from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis as QuadraticDiscriminantAnalysis
+#from sklearn.neighbors import KNeighborsClassifier as KNeighborsClassifier
+#from sklearn.pipeline import Pipeline
+#from sklearn.decomposition import PCA
+#from sklearn.tree import DecisionTreeClassifier as DecisionTreeClassifier
+
+DATASET_DIR = '/users/data/d3m/datasets/training_datasets'
+METADATA_PATH = '~/data/closed_d3m_unembed_data.csv'
+  
+def save_results(results,conf):
+    #save the results to a csv file
+    results.to_csv(model_name+'_final_cross_val.csv',index=False)
+    filename = model_name+'_matrix_mean.pkl'
+    fileObject = open(filename, 'wb')
+    pickle.dump(conf, fileObject)
+    fileObject.close()
+    
+def naive_gen():
+    model_name = 'Naive'
+    class NaiveModel:
+        def fit(self,X_train,y_train):
+            self.majority = y_train.value_counts().idxmax()
+        def predict(self,X_test):
+            y_hat = [self.majority for i in range(len(X_test))]
+            return y_hat
+    model = NaiveModel()
+    return model, model_name
+
+def evaluate_model(balance: bool, col_name: bool, use_metadata: bool, rank=None):
```

Comment:
Can you break up this function into smaller functions?

Reply:
Thank you for the type annotations, that helps. I am not sure what `col_name` is based on its name or how it is used below (see other comment). I think it is supposed to indicate whether we are using the column name only or using the dataset name and desc as well. If so, let's name it `use_col_name_only`.

Reply:
Is it possible to make the model and argument to this function too?